vktrace/src/vktrace_layer/vktrace_lib_trace.cpp

```diff
@@ -136,6 +138,161 @@ void *strip_create_extensions(const void *pNext)
+
+static std::list<void *> sighAddrList;
+static vktrace_sem_id sighAddrListSem = NULL;
+
+static void segvHandler(int sig, siginfo_t *si, void *ununsed)
+{
+    size_t pageSize = getpagesize();
+
+    // Note: we use sem_wait and mprotect inside a signal handler.
+    // This is not POSIX compliant, but works on Linux.
+
+    // Add the addr that caused the segv to sighAddrList
+    vktrace_sem_wait(sighAddrListSem);
+    sighAddrList.emplace_front(si->si_addr);
+    vktrace_sem_post(sighAddrListSem);
+
+    // Change protection of this page to allow the write to proceed
+    mprotect((void *)((uint64_t)si->si_addr & ~(pageSize-1)), pageSize, PROT_READ|PROT_WRITE);
+}
+
+
+// This function is called when we need to update our list of mapped memory
+// that is dirty.  On Linux, we use the /proc/self/pagemap to detect which pages
+// changed. But we also use mprotect with a signal handler for the rare case
+// in which mapped memory is written to between when we read the dirty bits
+// from /proc/self/pagemap dirty and clear the dirty bits.
+
+void getMappedDirtyPagesLinux(void)
+{
+    LPPageGuardMappedMemory pMappedMem;
+    PBYTE addr, alignedAddrStart, alignedAddrEnd;
+    uint64_t nPages;
+    VkDeviceMemory mappedMemory;
+    off_t pmOffset;
+    int64_t index;
+    size_t pageSize = getpagesize();
+    size_t readLen;
+    VKAllocInfo *pEntry;
+    struct sigaction sigAction;
+    static int pmFd = -1;
+    static std::vector<uint64_t> pageEntries;
+    
+    // If pageguard isn't enabled, we don't need to do anythhing
+    if (!getPageGuardEnableFlag())
+        return;
+
+    // Open pagefile, initialize sighAddrList semaphore, and set the SIGSEGV signal handler
+    if (pmFd == -1)
+    {
+        pmFd = open("/proc/self/pagemap", O_RDONLY);
+        if (pmFd < 0)
+            VKTRACE_FATAL_ERROR("Failed to open pagemap file.");
+
+        if (!vktrace_sem_create(&sighAddrListSem, 1))
+            VKTRACE_FATAL_ERROR("Failed to create sighAddrListSem.");
+
+        sigAction.sa_sigaction = segvHandler;
+        sigfillset(&sigAction.sa_mask);
+        sigAction.sa_flags = SA_SIGINFO;
+        if (0 != sigaction(SIGSEGV, &sigAction, NULL))
+            VKTRACE_FATAL_ERROR("sigaction sys call failed.");
+    }
+
+    // Iterate through all mapped memory allocations.
+    // Check dirty bits of each page in each mapped memory allocation.
+    for (std::unordered_map< VkDeviceMemory, PageGuardMappedMemory >::iterator it = getPageGuardControlInstance().getMapMemory().begin();
+         it != getPageGuardControlInstance().getMapMemory().end();
+         it++)
+    {
+        pMappedMem = &(it->second);
+        mappedMemory = pMappedMem->getMappedMemory();
+        pEntry=find_mem_info_entry(mappedMemory);
+        addr=pEntry->pData+pEntry->rangeOffset;
+        alignedAddrStart = (PBYTE)((uint64_t)addr & ~(pageSize-1));
+        alignedAddrEnd = (PBYTE)(((uint64_t)addr + pEntry->rangeSize + pageSize) & ~(pageSize-1));
```

Comment:
The calculation of alignedAddrEnd should add one less byte before masking.
(PBYTE)(((uint64_t)addr + pEntry->rangeSize + pageSize - 1) & ~(pageSize-1))
The for addr=0, pEntry->rangeSize = 4096, pageSize = 4096, alignedAddrEnd will be 4096 instead of 8192.
That will avoid mprotect giving ENOMEM for going one page past the end of a mapped region.